usr/sbin/pkcsslotd/slotmgr.c

```diff
@@ -329,182 +327,213 @@ static int create_pid_file(pid_t pid)
-                   sizeof(d->sinfo_struct.pk_slot.slotDescription), tok, val, ' '))
-            res = 1;
-        break;
-    case KW_MANUFID:
-        if (do_str(d, (char *)d->sinfo_struct.pk_slot.manufacturerID,
-                   sizeof(d->sinfo_struct.pk_slot.manufacturerID), tok, val, ' '))
-            res = 1;
-        break;
-    case KW_CONFNAME:
-        if (do_str(d, (char *)d->sinfo_struct.confname,
-                   sizeof(d->sinfo_struct.confname), tok, val, 0))
-            res = 1;
-        break;
-    case KW_TOKNAME:
-        if (do_str(d, (char *)d->sinfo_struct.tokname,
-                   sizeof(d->sinfo_struct.tokname), tok, val, 0))
-            res = 1;
-        break;
-    case KW_HWVERSION:
-        if (do_vers(d, &d->sinfo_struct.pk_slot.hardwareVersion, tok, val))
-            res = 1;
-        break;
-    case KW_FWVERSION:
-        if (do_vers(d, &d->sinfo_struct.pk_slot.firmwareVersion, tok, val))
-            res = 1;
-        break;
-    case KW_TOKVERSION:
-        if (do_vers(d, &vers, tok, val))
-            res = 1;
-        else
-            d->sinfo_struct.version = vers.major << 16 | vers.minor;
-        break;
-    default:
-        snprintf(d->errbuf, sizeof(d->errbuf),
-                 "Unknown string-valued keyword detected: \"%s\"",
-                 keyword_token_to_str(tok));
-        res = 1;
-        break;
+    ret = parse_configlib_file(file, &config, config_parse_error, 0);
+    fclose(file);
+    if (ret != 0) {
+        ErrLog("Error parsing config file '%s'\n", config_file);
+        goto done;
     }
-    return res;
-}
 
-static int slotmgr_key_vers(void *private, int tok, unsigned int vers)
-{
-    struct parse_data *d = (struct parse_data *)private;
+    confignode_foreach(c, config, i) {
+        DbgLog(DL3, "Config node: '%s' type: %u line: %u\n",
+               c->key, c->type, c->line);
 
-    switch (tok) {
-    case KW_TOKVERSION:
-        d->sinfo_struct.version = vers;
-        return 0;
-    case KW_HWVERSION:
-        d->sinfo_struct.pk_slot.hardwareVersion.major = vers >> 16;
-        d->sinfo_struct.pk_slot.hardwareVersion.minor = vers & 0xffu;
-        return 0;
-    case KW_FWVERSION:
-        d->sinfo_struct.pk_slot.firmwareVersion.major = vers >> 16;
-        d->sinfo_struct.pk_slot.firmwareVersion.minor = vers & 0xffu;
-        return 0;
-    }
-    snprintf(d->errbuf, sizeof(d->errbuf),
-             "Unkown version-valued keyword detected: \"%s\"",
-             keyword_token_to_str(tok));
-    return 1;
-}
+        if (confignode_hastype(c, CT_FILEVERSION)) {
+            DbgLog(DL0, "Config file version: '%s'\n",
+                   confignode_to_fileversion(c)->base.key);
+            continue;
+        }
 
-static void slotmgr_disab_event_supp(void *private)
-{
-    UNUSED(private);
+        if (confignode_hastype(c, CT_BARECONST)) {
+            if (strcmp(confignode_to_bareconst(c)->base.key,
```

Comment:
You could simply use `c->key` here since `&(confignode_to_bareconst(c)->base) == c`.

Reply:
yes I know, but to be totally implementation independent, I decided to do it that way :-)